pkg/services/cleanup/cleanup.go

```diff
@@ -57,66 +62,97 @@ type CleanUpService struct {
-			srv.expireOldUserInvites(ctx)
-			srv.deleteStaleShortURLs(ctx)
-			srv.deleteStaleQueryHistory(ctx)
-			err := srv.ServerLockService.LockAndExecute(ctx, "delete old login attempts",
-				time.Minute*10, func(context.Context) {
-					srv.deleteOldLoginAttempts(ctx)
-				})
-			if err != nil {
-				srv.log.Error("failed to lock and execute cleanup of old login attempts", "error", err)
-			}
+			srv.clean(ctx)
 		case <-ctx.Done():
 			return ctx.Err()
 		}
 	}
 }
 
-func (srv *CleanUpService) cleanUpOldAnnotations(ctx context.Context) {
+func (srv *CleanUpService) clean(ctx context.Context) {
+	const timeout = 9 * time.Minute
+	start := time.Now()
+	ctx, span := srv.tracer.Start(ctx, "cleanup background job")
+	defer span.End()
+	tID := tracing.TraceIDFromContext(ctx, false)
+	logger := srv.log.New("traceID", tID)
+	ctx, cancelFn := context.WithTimeout(ctx, timeout)
+	defer cancelFn()
+
+	cleanupJobs := []cleanUpJob{
+		{"clean up temporary files", srv.cleanUpTmpFiles},
+		{"delete expired snapshots", srv.deleteExpiredSnapshots},
+		{"delete expired dashboard versions", srv.deleteExpiredDashboardVersions},
+		{"delete expired images", srv.deleteExpiredImages},
+		{"cleanup old annotations", srv.cleanUpOldAnnotations},
+		{"expire old user invites", srv.expireOldUserInvites},
+		{"delete stale short URLs", srv.deleteStaleShortURLs},
+		{"delete stale query history", srv.deleteStaleQueryHistory},
+		{"delete old login attempts", srv.deleteOldLoginAttempts},
+	}
+
+	logger.Debug("Starting cleanup jobs", "jobs", fmt.Sprintf("%v", cleanupJobs))
+
+	for _, j := range cleanupJobs {
+		if ctx.Err() != nil {
+			logger.Error("Cancelled cleanup job", "error", ctx.Err(), "duration", time.Since(start))
+			return
+		}
+		ctx, span := srv.tracer.Start(ctx, j.name)
+		j.fn(ctx, logger)
+		span.End()
+	}
+
+	logger.Info("Completed cleanup jobs", "duration", time.Since(start))
+}
+
+func (srv *CleanUpService) cleanUpOldAnnotations(ctx context.Context, logger log.Logger) {
 	cleaner := annotations.GetAnnotationCleaner()
 	affected, affectedTags, err := cleaner.CleanAnnotations(ctx, srv.Cfg)
 	if err != nil && !errors.Is(err, context.DeadlineExceeded) {
-		srv.log.Error("failed to clean up old annotations", "error", err)
+		logger.Error("failed to clean up old annotations", "error", err)
 	} else {
-		srv.log.Debug("Deleted excess annotations", "annotations affected", affected, "annotation tags affected", affectedTags)
+		logger.Debug("Deleted excess annotations", "annotations affected", affected, "annotation tags affected", affectedTags)
 	}
 }
 
-func (srv *CleanUpService) cleanUpTmpFiles() {
+func (srv *CleanUpService) cleanUpTmpFiles(ctx context.Context, logger log.Logger) {
 	folders := []string{
 		srv.Cfg.ImagesDir,
 		srv.Cfg.CSVsDir,
 	}
 
 	for _, f := range folders {
-		srv.cleanUpTmpFolder(f)
+		ctx, span := srv.tracer.Start(ctx, fmt.Sprintf("removing old temporary items from directory: %s", strconv.Quote(f)))
```

Comment:
Nit. This is a question regarding best practices related to naming spans and high cardinality. Based on what I've read span names should not include variable parameters. In this example it can have 2 variations which I guess is fine and therefore the nit.

Can't find official opentelemetry docs about it right now which is weird, but for example https://uptrace.dev/opentelemetry/span-naming.html#span-names is a good source.

Reply:
I was wondering about naming vs labels.

I’m not sure if that’s directly applicable here. Their guidance seem to be similar to Prometheus labels: don’t let users affect the cardinality. They do suggest using paths and queries without user input as names, which is a higher cardinality than cleanup jobs. Though it seem to imply that I should aim at keeping names shorter, at least.

I’ll give this another pass.